File: typescript/src/builders/compose-builder.ts
import type { DocumentsClient } from "../clients/documents.js";
import type {
  ContentArea,
  StationeryConfig,
  WatermarkConfig,
} from "../models/content.js";
import type { ComposeRequest, DocumentPart } from "../models/document.js";
import type { OutputFormat } from "../models/enums.js";
import type { InvoiceData } from "../models/invoice.js";
import { toBase64, type FileInput } from "../files.js";

/**
 * Fluent builder for multi-part document composition.
 *
 * @example
 * ```ts
 * const pdf = await dg.compose()
 *   .part({ htmlContent: '<h1>Cover</h1>' })
 *   .part({ htmlContent: '<h1>Chapter 1</h1>', templateName: 'report.odt' })
 *   .watermark('CONFIDENTIAL')
 *   .asPdf()
 *   .generate();
 * ```
 */
export class ComposeBuilder {
  private readonly request: ComposeRequest = { parts: [] };
  private client?: DocumentsClient;

  /** @internal Attach a client for direct generation. */
  _setClient(client: DocumentsClient): this {
    this.client = client;
    return this;
  }

  /** Add a document part. */
  part(p: DocumentPart): this {
    this.request.parts.push(p);
    return this;
  }

  /** Add a diagonal watermark overlay. */
  watermark(text: string, options?: Omit<WatermarkConfig, "text">): this {
    if (options) {
      this.request.watermark = { text, ...options };
    } else {
      this.request.watermark = text;
    }
    return this;
  }

  /** Set stationery (letterhead) PDF background. */
  stationery(file: FileInput, firstPageFile?: FileInput): this {
    const config: StationeryConfig = { pdfBase64: toBase64(file) };
    if (firstPageFile) {
      config.firstPagePdfBase64 = toBase64(firstPageFile);
    }
    this.request.stationery = config;
    return this;
  }

  /** Add an absolutely positioned content area. */
  contentArea(area: ContentArea): this {
    this.request.contentAreas ??= [];
    this.request.contentAreas.push(area);
    return this;
  }

  /** Attach structured invoice data. */
  invoice(data: InvoiceData): this {
    this.request.invoiceData = data;
    return this;
  }

  /** Set a password for the output document. */
  password(pwd: string): this {
    this.request.password = pwd;
    return this;
  }

  /** Set the output format. */
  outputFormat(format: OutputFormat): this {
    this.request.outputFormat = format;
    return this;
  }

  /** Shorthand: set output to PDF. */
  asPdf(): this {
    this.request.outputFormat = "PDF";
    return this;
  }

  /** Shorthand: set output to DOCX. */
  asDocx(): this {
    this.request.outputFormat = "DOCX";
    return this;
  }

  /** Build the ComposeRequest without executing. */
  build(): ComposeRequest {
    return { ...this.request, parts: [...this.request.parts] };
  }

  /** Generate the composed document (requires attached client). */
  async generate(): Promise<Buffer> {
    if (!this.client) {
      throw new Error(
        "No client attached. Use dg.compose() or pass the request to dg.documents.compose().",
      );
    }
    return this.client.compose(this.build());
  }
}

File: java/src/main/java/com/dokmatiq/docgen/builder/ComposeBuilder.java
package com.dokmatiq.docgen.builder;

import com.dokmatiq.docgen.client.DocumentsClient;
import com.dokmatiq.docgen.internal.FileUtils;
import com.dokmatiq.docgen.model.*;

import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

/**
 * Fluent builder for multi-part document composition.
 *
 * <pre>{@code
 * byte[] pdf = dg.compose()
 *     .part(new DocumentPart("<h1>Cover</h1>"))
 *     .part(new DocumentPart("<h1>Chapter 1</h1>", "report.odt"))
 *     .watermark("CONFIDENTIAL")
 *     .asPdf()
 *     .generate();
 * }</pre>
 */
public class ComposeBuilder {
    private DocumentsClient client;
    private final List<DocumentPart> parts = new ArrayList<>();
    private Object watermark;
    private StationeryConfig stationery;
    private List<ContentArea> contentAreas;
    private InvoiceData invoiceData;
    private String password;
    private OutputFormat outputFormat;
    private String callbackUrl;
    private String callbackSecret;

    /** @internal */
    public ComposeBuilder setClient(DocumentsClient client) {
        this.client = client;
        return this;
    }

    public ComposeBuilder part(DocumentPart part) { parts.add(part); return this; }

    public ComposeBuilder watermark(String text) { this.watermark = text; return this; }
    public ComposeBuilder watermark(WatermarkConfig config) { this.watermark = config; return this; }

    public ComposeBuilder stationery(Path file) {
        this.stationery = new StationeryConfig(FileUtils.toBase64(file), null);
        return this;
    }

    public ComposeBuilder contentArea(ContentArea area) {
        if (contentAreas == null) contentAreas = new ArrayList<>();
        contentAreas.add(area);
        return this;
    }

    public ComposeBuilder invoice(InvoiceData data) { this.invoiceData = data; return this; }
    public ComposeBuilder password(String pwd) { this.password = pwd; return this; }
    public ComposeBuilder outputFormat(OutputFormat format) { this.outputFormat = format; return this; }
    public ComposeBuilder asPdf() { this.outputFormat = OutputFormat.PDF; return this; }
    public ComposeBuilder asDocx() { this.outputFormat = OutputFormat.DOCX; return this; }

    public ComposeBuilder callback(String url, String secret) {
        this.callbackUrl = url;
        this.callbackSecret = secret;
        return this;
    }

    /** Build the ComposeRequest without executing. */
    public ComposeRequest build() {
        return new ComposeRequest(List.copyOf(parts), watermark, stationery, contentAreas,
                invoiceData, password, outputFormat, callbackUrl, callbackSecret);
    }

    /** Generate the composed document (requires attached client). */
    public byte[] generate() {
        if (client == null) {
            throw new IllegalStateException(
                    "No client attached. Use dg.compose() or pass the request to dg.documents().compose().");
        }
        return client.compose(build());
    }
}
